### PR TITLE
Custom Frequency Picker

### PR DESCRIPTION
**Pull request by suhaibabsi-inst**
refs: MBL-17525
affects: Student
release note: Frequency selector for Calendar Event creation.

## What's new
- Introducing screen to allow for selecting Event frequency out of predefined presets.
- Introducing screen to allow build a custom Event frequency rule.

### Chores
- Introducing `InstUI.Icons` for functions like back, picking, checking.
- Introducing a new `MultiPickerView` to allow multi-component picking. 
Used for (Interval-Frequency) picker at the top.  
- Introducing `DropDownButton` & `DropDownDetailsViewModifier` to allow building dynamic dropDowns with custom controls, as an example see `WeekDaysSelectionListView`, which used in `EditCustomFrequencyScreen.weekDaysCell`.
- Introducing `SelectionMenu` to allow for the case of "no selection". SwiftUI.Picker doesn't support that
- Added few utils functions to Array, Number, String and Date extensions.

## What's not included
- Weekdays is assumed to be (Monday to Friday), localization for it is still to be decided.
- Updating a predefined recurring event frequency rule.
- Advanced validation against end-date, by which number of occurrences is calculated to the selected end date, then validated against (400 max.) rule.

## Test plan
- Smoke test creating event with preset frequency .
- Smoke test creating event with a custom frequency.
- Smoke test editing the frequency rule of recurring even. 

## Screenshots

<table>
<tr>
<td></td>
<td></td>
</tr>
<tr>
<td></td>
<td></td>
</tr>
<tr>
<td></td>
<td></td>
</tr>
<tr>
<td colspan="2" ></td>
</tr>
<tr>
<td colspan="2" ></td>
</tr>
<tr>
<td colspan="2"></td>
</tr>
</table>

## Checklist

- [ ] Follow-up e2e test ticket created
- [x] A11y checked
- [x] Tested on phone
- [x] Tested on tablet
- [x] Tested in dark mode
- [x] Tested in light mode
- [ ] Approve from product

### CODE REVIEW

**Comment by inst-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>Build failed, skipping coverage check</td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>This pull request will not generate a release note.</td>
    </tr>
  </tbody>
</table>


#### Affected Apps: Student

[MBL-17293](https://instructure.atlassian.net/browse/MBL-17293)

<details><summary>:x: Swift lint</summary>


    ❌ /Users/vagrant/git/Core/Core/InstUI/Views/Cells/DropDownCell.swift:58:30: Colon Spacing Violation: Colons should be next to the identifier when specifying a type and next to the key in dictionary literals (colon)
    /Users/vagrant/git/Core/Core/InstUI/Views/NavigationBarButton.swift:193:1: error: Trailing Whitespace Violation: Lines should not have trailing whitespace (trailing_whitespace)
    /Users/vagrant/git/Core/Core/Planner/CalendarEvent/ViewModel/EditCalendarEventViewModel.swift:286:1: error: Trailing Whitespace Violation: Lines should not have trailing whitespace (trailing_whitespace)
    ❌ /Users/vagrant/git/Core/Core/Planner/CalendarEvent/Model/UseCase/RecurrenceRule.swift:213:1: Trailing Whitespace Violation: Lines should not have trailing whitespace (trailing_whitespace)
    /Users/vagrant/git/Core/Core/Planner/CalendarEvent/View/EditCustomFrequencyScreen.swift:57:85: error: Unused Closure Parameter Violation: Unused parameter in a closure should be replaced with _ (unused_closure_parameter)
    
</details>
<details><summary>:x: XCTest failed: CoreTests/APICalendarEventRequestBodyTests/testEncodingShouldNotSkipNils</summary>


    XCTAssertEqual failed: ("false") is not equal to ("true")
    XCTAssertEqual failed: ("false") is not equal to ("true")
    XCTAssertEqual failed: ("false") is not equal to ("true")
    XCTAssertEqual failed: ("false") is not equal to ("true")
</details>

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 53227b714adda2a9d778d61a8868d1c7bfda205c
</p>

**Comment by inst-danger** [comment]
<details><summary>Student Build QR Code:</summary>

</details>

**Review by vargaat — CHANGES_REQUESTED** [review]
Good job on implementing all those, this wasn't an easy task! I checked the code and quickly checked on the UI and the behavior and had some findings:

- Please accept the CLA.
- Add more tests to new entities to reach the minimum code coverage. Check the coverage table in the PR comment.

- The values on the day picker are not visible in dark mode.
- The picker's right side goes out of the modal and cut off.


- I could enter an invalid value here.
- There's no OK in the dialog but I guess this is just because I could enter an invalid amount there.
- Not visible that well on this screenshot but the up and down arrows at the end of the row are not really visible in dark mode.

File: Core/Core/InstUI/Styles/Padding.swift
[the remark this answers is not in the document]
Comment (suhaibabsi-inst) [reply Core/Core/InstUI/Styles/Padding.swift:1]: This is to allow to define multiple cases that could have the same value.

File: Core/Core/InstUI/Views/DropDown/DropDownButton.swift [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+
+extension Array where Element: Equatable {
+
+    mutating func toggleInsert(with value: Element) {
+        if contains(value) {
+            removeAll(where: { $0 == value })
+        } else {
+            append(value)
+        }
+    }
```
[the remark this answers is not in the document]
Comment (szabinst) [reply Core/Core/InstUI/Views/DropDown/DropDownButton.swift:178]: And I'd move it to an extension file.

**Review by szabinst — COMMENTED** [review]
Congrats on your first PR! 

My initial findings while trying out the feature:
- When I first select a day on `Repeats on` the row’s position changes. I’d assume it’s because “Not selected” without background is smaller in height than e.g “Monday”. 
- Sometimes I wasn’t able to select “Monday” on the `Repeats on` picker because it dismissed the picker. 
- I wasn’t able to create a recurring event with custom frequency
- Minor inconvenience but setting up an event with incorrect end date e.g  with 2 week, Fri-Sat-Sun repeat, ending on September 3 gives an error only when I try to submit the event. Maybe we could add some error handling before like preventing the selection of an end date that doesn’t make sense. 
- On the wheel picker when I start changing the number it’s not always synchronized with the plural form of “day”, “week” etc. 

This is more a UX thing but for me the wheel picker with 1, 2, 3 / days, weeks wasn’t really intuitive.

File: Core/Core/InstUI/Views/NavigationBarButton.swift [hunk 2]
```diff
@@ -179,6 +179,28 @@ extension InstUI.NavigationBarButton {
...
+                AnyView(
+                    HStack(spacing: 5) {
+                        InstUI.Icons.Back().foregroundStyle(Color.textDarkest)
+                        Text("Back", bundle: .core)
+                    }
+                )
+            }
+        )
+    }
+
```
[the remark this answers is not in the document]
Comment (suhaibabsi-inst) [reply Core/Core/InstUI/Views/NavigationBarButton.swift:203]: The reason for adding that, is to provide the option to trigger a custom action just before dismissing the view.

[the remark this answers is not in the document]
Comment (vargaat) [reply Core/Core/InstUI/Views/NavigationBarButton.swift:203]: Could we use the `.onDisappear` view modifier to notify the view model about the screen being dismissed?

**Review by vargaat — CHANGES_REQUESTED** [review]
- I would expect that pressing `Done` here will pop me back to the event editor screen because I have finished editing the frequency. Instead it pops back to the pre-defined list of events where I have to press back again to get back to the event itself.
- Another thing that we could improve is to center the picker wheel's content horizontally. The selection indicator itself is centered but within it the numbers and the text are left aligned. If it's hard to do we can leave it as-is.

**Comment by suhaibabsi-inst** [comment]
Thanks @szabinst for the attentive feedback. 

> Congrats on your first PR!
> 
> My initial findings while trying out the feature:
> 
> * When I first select a day on `Repeats on` the row’s position changes. I’d assume it’s because “Not selected” without background is smaller in height than e.g “Monday”.

**Fixed.**

> * Sometimes I wasn’t able to select “Monday” on the `Repeats on` picker because it dismissed the picker.

**Fixed.**

> * I wasn’t able to create a recurring event with custom frequency

**Can you pass a demo record for that failing use case? Perhaps, I can use the same account for checking on my end.**

> * Minor inconvenience but setting up an event with incorrect end date e.g  with 2 week, Fri-Sat-Sun repeat, ending on September 3 gives an error only when I try to submit the event. Maybe we could add some error handling before like preventing the selection of an end date that doesn’t make sense.

**We can match what we currently have on web. which is only checks if the selected date is larger than eventDate, and that total count of possible occurrences to that date is less than 400.**

**For this PR, the part about checking if selected date is larger than or equal to event date was implemented.
For the second part, I think we might need to put more effort into it. It can be a bit of a complex logic.**

> * On the wheel picker when I start changing the number it’s not always synchronized with the plural form of “day”, “week” etc.

**I made few fixes to this UI component. Can you pass in a demo record for that failing use case?** 

> 
> This is more a UX thing but for me the wheel picker with 1, 2, 3 / days, weeks wasn’t really intuitive.

**Perhaps, we can use Marcell's feedback. Will reach out to him on that.**

**Review by rh12 — CHANGES_REQUESTED** [review]
Thanks for making the deep dive into all this!
I have some findings and also some comments on code (nitpicking a lot, since this is a first PR to get more familiar with implicit "rules" we have)

UI changes to followup (discussed in private):
- CustomFrequency screen: "Repeats every" label at the top (cc @szabinst, hopefully it helps making the picker make more sense)
- CustomFrequency screen: Divider under the picker

UI findings:
- according to Figma (doublechecked with Marci), the "End Repeat" dropdown [menu](https://www.figma.com/design/asf3Fh7xqLsFpVOJND5QLd/%E2%9C%85-%5BDelivered%5D-(Canvas-Teacher)-Calendar-V01--Marci?node-id=7425-13499&t=Bo3Rqgr1idre5DWY-4) should be the same style as the weekday selector [dropdown](https://www.figma.com/design/asf3Fh7xqLsFpVOJND5QLd/%E2%9C%85-%5BDelivered%5D-(Canvas-Teacher)-Calendar-V01--Marci?node-id=7425-13531&t=Bo3Rqgr1idre5DWY-4) for "Repeats on".
(This may result in SelectionMenu not being needed at all, but I'm still figuring out that part)

Will continue with review

File: Core/Core/InstUI/Views/NavigationBarButton.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (rh12) [reply Core/Core/InstUI/Views/NavigationBarButton.swift]: Unresolving. It seems to we still use the custom back buttons.

File: Core/Core/Planner/CalendarEvent/ViewModel/EditCalendarEventViewModel.swift [hunk 3]
```diff
@@ -45,8 +37,8 @@ final class EditCalendarEventViewModel: ObservableObject {
     @Published var isAllDay: Bool = false
     @Published var startTime: Date?
     @Published var endTime: Date?
-    @Published private(set) var frequencyName: String?
-    @Published private(set) var calendarName: String?
+    @Published var frequency: FrequencySelection?
+    @Published var calendarName: String?
```
Comment (rh12) [inline Core/Core/Planner/CalendarEvent/ViewModel/EditCalendarEventViewModel.swift:41]: this could be reverted to `private(set)`, as it was

File: Core/Core/InstUI/Views/NavigationBarButton.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (suhaibabsi-inst) [reply Core/Core/InstUI/Views/NavigationBarButton.swift]: @rh12 
This is strange. This part is already removed from code.
If you go to tab "Files Changed", you won't see it there.

[the remark this answers is not in the document]
Comment (rh12) [reply Core/Core/InstUI/Views/NavigationBarButton.swift]: Sorry, my bad. Probably I've tried it on an earlier commit.
Could you please remove `InstUI.Icons.Back` in that case?

**Review by rh12 — CHANGES_REQUESTED** [review]

File: Core/Core/Planner/CalendarEvent/Model/Helpers/RecurrenceRule.swift [hunk 4]
```diff
@@ -0,0 +1,456 @@
...
+    init?(component: Int) {
+        guard let first = Self
+            .allCases
+            .first(where: { $0.dateComponent == component })
+        else { return nil }
+        self = first
+    }
+}
+
+struct DayOfWeek: Equatable, RRuleCodable {
```
[the remark this answers is not in the document]
Comment (suhaibabsi-inst) [reply Core/Core/Planner/CalendarEvent/Model/Helpers/RecurrenceRule.swift:225]: @rh12 

While I agree with you about the type of `DayOfMonth`.  I should say 
Types `DayOfMonth`, `DayOfYear` & `RecurrenceEndMode` are types that are invented to serve selection controls on screen `EditCustomFrequencyScreen`. So I am going to nest those into `EditCustomFrequencyViewModel`.

While for the type `DayOfWeek`, it serve a much broader context, which is relevant to initializing and tweaking instances of `RecurrenceRule`. Therefore, will be nesting it into that type. 

As personal preference, generally I wouldn't nest a type unless I came across a necessity that forces me to do so. Would rather try to be more creative when defining new types. :)

File: Core/Core/Planner/CalendarEvent/ViewModel/EditCustomFrequencyViewModel.swift [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+
+    @Published var frequency: RecurrenceFrequency = .daily
+    @Published var interval = FrequencyInterval(value: 1)
+
+    @Published var endMode: EndMode?
+    @Published var endDate: Date? = Clock.now
+    @Published var occurrenceCount: Int
+
+    /// Specific to Weekly
+    @Published var daysOfTheWeek: [Weekday] = []
```
Comment (rh12) [inline Core/Core/Planner/CalendarEvent/ViewModel/EditCustomFrequencyViewModel.swift:49]: We seems to have missed this, but this should default to the currently selected day.
It was already in the figma, but now it's [highlighted](https://www.figma.com/design/asf3Fh7xqLsFpVOJND5QLd/%E2%9C%85-%5BDelivered%5D-(Canvas-Teacher)-Calendar-V01--Marci?node-id=8008-13658&t=wbEuSvUtdNY7YJCQ-4).

File: Core/Core/Planner/CalendarEvent/ViewModel/EditCustomFrequencyViewModel.swift [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+
+    /// Specific to Monthly
+    @Published var dayOfMonth: DayOfMonth?
```
Comment (rh12) [inline Core/Core/Planner/CalendarEvent/ViewModel/EditCustomFrequencyViewModel.swift:52]: We seems to have missed this, but this should default to the currently selected day of month.
It was already in the figma, but now it's [highlighted](https://www.figma.com/design/asf3Fh7xqLsFpVOJND5QLd/%E2%9C%85-%5BDelivered%5D-(Canvas-Teacher)-Calendar-V01--Marci?node-id=8012-27702&t=wbEuSvUtdNY7YJCQ-4).
This can't be unselected, so it may make sense to make it non-optional, if it doesn't complicate init too much

File: Core/Core/Planner/CalendarEvent/ViewModel/EditCustomFrequencyViewModel.swift [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+            }
+
+        } else {
+            let fewDays = weekdays.count < 3
+            tags.append(contentsOf: weekdays.map { wday in
+                return fewDays ? wday.pluralText : wday.shortText
+            })
+        }
+
+        return tags.nilIfEmpty ?? [String(localized: "Not selected", bundle: .core)]
```
Comment (rh12) [inline Core/Core/Planner/CalendarEvent/ViewModel/EditCustomFrequencyViewModel.swift:240]: I'm not sure, but aren't we ignoring this string because we display the PromptLabel instead of the LabelButtons?

**Review by rh12 — CHANGES_REQUESTED** [review]

File: Core/Core/Planner/CalendarEvent/View/EditCustomFrequencyScreen.swift [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+            trailing: .init(
+                isEnabled: viewModel.isSaveButtonEnabled,
+                isAvailableOffline: false,
+                title: viewModel.doneButtonTitle,
+                action: {
+                    viewModel.didTapDone.send(viewController)
+                }
+            )
+        )
+        .dropDownDetails(state: $weekDayDropDownState) {
```
[the remark this answers is not in the document]
Comment (suhaibabsi-inst) [reply Core/Core/Planner/CalendarEvent/View/EditCustomFrequencyScreen.swift:86]: This modifier should be used on the target container where the dropDown details is to be displayed.
Will do to some renaming to make more clear, and perhaps add some comments.

File: Core/Core/Planner/CalendarEvent/View/CustomFrequencyComponents/WeekDaysSelectionListView.swift [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+//
+
+import SwiftUI
+
+struct WeekDaysSelectionListView: View {
```
[the remark this answers is not in the document]
Comment (suhaibabsi-inst) [reply Core/Core/Planner/CalendarEvent/View/CustomFrequencyComponents/WeekDaysSelectionListView.swift:21]: Event if not used in the `var body: some View { .. }` ?
I checked in code whether this is being used in some way, but couldn't figure it out. 
Can you elaborate more about that ?

File: Core/Core/Planner/CalendarEvent/View/EditCustomFrequencyScreen.swift [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+        return InstUI.LabelValueCell(
+            label: Text("Repeats on", bundle: .core),
+            value: viewModel.titleForProposedDayOfYear,
+            equalWidth: false,
+            action: {}
+        )
+    }
+
+    private var weekDaysCell: some View {
+        InstUI.DropDownCell(
```
[the remark this answers is not in the document]
Comment (suhaibabsi-inst) [reply Core/Core/Planner/CalendarEvent/View/EditCustomFrequencyScreen.swift:113]: @rh12 

We can discuss that more on live meeting. But here some little thoughts about it:
Native menus by design has a lot to offer than a custom built one, except for styling. More accessible, and familiarity (user can expect to do), compatibility with system, and accessibility. 

So my take would be to use both ways:
We can use our custom built DropDown component for versions below iOS 16.4.
And we can use the native version as an upgraded experience for iOS 16.4+, given that came to a resolution to the issue of fading buttons on selection.

File: Core/Core/Planner/CalendarEvent/View/CustomFrequencyComponents/OccurrencesCountInputDialog.swift [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+                    Text("Cancel", bundle: .core)
+                })
+
+                Button(action: {
+                    inputModel.submit()
+                    isPresented = false
+                }, label: {
+                    Text("Done", bundle: .core)
+                })
+                .disabled(inputModel.isValid == false)
```
[the remark this answers is not in the document]
Comment (suhaibabsi-inst) [reply Core/Core/Planner/CalendarEvent/View/CustomFrequencyComponents/OccurrencesCountInputDialog.swift:55]: Disable behavior for Alert's Actions seems to work only for iOS 17+.
And for it work properly, we need to make sure it is not disabled when onAppear. 
I made few changes on it to get around this limitation. I had to revert back to use `text` binding of `TextField`.
Which is not a thing I was hoping for 😄

File: Core/Core/Planner/CalendarEvent/View/CustomFrequencyComponents/WeekDaysSelectionListView.swift (hunk 9 above)
[the remark this answers is not in the document]
Comment (rh12) [reply Core/Core/Planner/CalendarEvent/View/CustomFrequencyComponents/WeekDaysSelectionListView.swift:21]: I think we rarely (if ever) need to use the actual value. The point is, that since this is an `@Environment` property, when it changes (when the user changes text size) the view which contains it will be redrawn, updating the text size instantly.

File: Core/Core/Planner/CalendarEvent/View/CustomFrequencyComponents/OccurrencesCountInputDialog.swift (hunk 11 above)
[the remark this answers is not in the document]
Comment (rh12) [reply Core/Core/Planner/CalendarEvent/View/CustomFrequencyComponents/OccurrencesCountInputDialog.swift]: Thank you! It seems to work fine on iOS17!
On iOS16 if I input something invalid and press Done (as you mentioned, disable is not working pre 17) it will fallback to some value (the first digit of the previous value). My only idea here for iOS16, if it's a number: clamp it. If it's not, fallback to 0. (so the user can't tap Done with some unintended value)

**Review by rh12 — CHANGES_REQUESTED** [review]
Thanks for all the code review changes!

I've started QA-ing the app and here are some findings.
I will check VoiceOver after this one, but that should be the end of it.

File: Brewfile.lock.json
[the remark this answers is not in the document]
Comment (szabinst) [reply Brewfile.lock.json:1]: This is caused by running `make provision`. I think we can all update brew.

File: Core/Core/InstUI/Views/MultiPickerView.swift
[the remark this answers is not in the document]
Comment (szabinst) [reply Core/Core/InstUI/Views/MultiPickerView.swift:1]: I guess this could be caused by having different english (UK, US) on your physical device and on simulator.

File: Core/Core/Planner/CalendarEvent/View/EditEventFrequencyScreen.swift [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+//
+
+import SwiftUI
+
+struct EditEventFrequencyScreen: View, ScreenViewTrackable {
```
Comment (rh12) [inline Core/Core/Planner/CalendarEvent/View/EditEventFrequencyScreen.swift:21]: Please add Selected to VoiceOver for the selected row, and also read out when it get's selected, as an audible feedback. You can check this in "Teacher app > Course > Modules > tap the published icon for a File" (it is `ModuleFilePermissionEditorView` in code)

**Review by rh12 — CHANGES_REQUESTED** [review]
IMHO the MultiPicker is a bit hard to follow in VoiceOver, but improving that is probably too much for now.
Besides that, I only encountered one issue regarding VoiceOver. All else sounds as expected I think!

File: Core/Core/Planner/CalendarEvent/View/EditCustomFrequencyScreen.swift [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+                isEnabled: viewModel.isSaveButtonEnabled,
+                isAvailableOffline: false,
+                title: viewModel.doneButtonTitle,
+                action: {
+                    viewModel.didTapDone.send(viewController)
+                }
+            )
+        )
+        .dropDownDetailsContainer(state: $weekDayDropDownState) {
+            WeekDaysSelectionListView(selection: $viewModel.daysOfTheWeek)
```
[the remark this answers is not in the document]
Comment (suhaibabsi-inst) [reply Core/Core/Planner/CalendarEvent/View/EditCustomFrequencyScreen.swift:87]: Fixed. There should be no interaction going when list is presented.

**Review by rh12 — DISMISSED** [review]
QA + 1

**Comment by ndrsszsz** [comment]
#### I think "On date" should be "On Date" to be consistent with other places in the app.</span>



#### I don't think 0 occurences should be accepted in this case (but I'm not sure about this one).




#### Plurals are not handled correctly here. See the video for details.

https://github.com/user-attachments/assets/6c950bc7-f37b-4077-a899-f48b71d8cf20

**Review by rh12 — APPROVED** [review]

**Review by ndrsszsz — APPROVED** [review]
QA +1

**Review by vargaat — APPROVED** [review]
